### PR TITLE
Remove subscription

### PR DESCRIPTION
**Pull request by BrettRD**
This allows subscriptions (and other callbacks) to be removed from the executor
It's been a while since I worked on this, I'll mark it as draft and re-base the commit
edit: fast-forward was pretty clean

### CODE REVIEW

**Comment by BrettRD** [comment]
This is ready for review

**Review by JanStaschulat — CHANGES_REQUESTED** [review]
Source code looks great to me. Some remarks: 
- By copying the values, the order of handels is preserved. That is good (because this defines the processing order of callbacks in the the spin methods). 
- You initialize the handle (which was copied to the n-1th element, good!
- test case, handle_list_size=1, remove one subscription
- memory of the handle is not freed because it is allocated elsewhere (in main), OK

Could you please add some more test cases for removing subscriptions:
- handle_list: [a,b,c], remove a
- handle_list: [a,b,c], remove b
- handle_list: [a,b,c], remove c

Could you add one test case each for 
- removing timer
- guard_condition
- service
- client

correct DCO errors as described in 'Details' https://github.com/ros2/rclc/pull/106/checks?check_run_id=2948178382

**Comment by JanStaschulat** [comment]
I had to update the build jobs for foxy branch: 
- updated ci job (GPG incident) 
- removed Dashing from build jobs 
in this PR: https://github.com/ros2/rclc/pull/110

once merged into branch foxy, please pull and merge into your PRs.

**Comment by BrettRD** [comment]
New tests: removal of handles at every meaningful position in the executor handle list and their effect on list order, API tests for every other handle type.

I jumped the gun on your CI fix and I'm getting a merge conflict doing a bulk-sign-off, so I'll wait for https://github.com/ros2/rclc/pull/110 to hit foxy before sorting that out.

File: rclc/test/rclc/test_executor.cpp
Comment (JanStaschulat) [inline rclc/test/rclc/test_executor.cpp:812]:
```suggestion
  EXPECT_EQ(executor.handles[0].subscription, &this->sub1) <<
    "expect to find sub1 in first handle";
  EXPECT_EQ(executor.handles[1].subscription, &this->sub2) <<
    "expect to find sub2 in second handle";
```

File: rclc/test/rclc/test_executor.cpp
Comment (JanStaschulat) [inline rclc/test/rclc/test_executor.cpp:836]:
```suggestion
  EXPECT_EQ(executor.handles[0].subscription, &this->sub1) <<
    "expect to find sub1 in first handle";
  EXPECT_EQ(executor.handles[1].subscription, &this->sub3) <<
    "expect to find sub3 in second handle";
```

File: rclc/test/rclc/test_executor.cpp
Comment (JanStaschulat) [inline rclc/test/rclc/test_executor.cpp:860]:
```suggestion
  EXPECT_EQ(executor.handles[0].subscription, &this->sub3) <<
    "expect to find sub3 in first handle";
  EXPECT_EQ(executor.handles[1].subscription, &this->sub2) <<
    "expect to find sub2 in second handle";
```

File: rclc/test/rclc/test_executor.cpp
Comment (JanStaschulat) [reply rclc/test/rclc/test_executor.cpp:812]: I added this test directly after the remove-operation folowing the principle: one method-call (here `rclc_executor_remove_subscription`) then test expected value with actual value. I saw, of course, that you test the same after adding the subscription again at the end.

**Review by JanStaschulat — CHANGES_REQUESTED** [review]

**Comment by JanStaschulat** [comment]
@BrettRD PR #110 is merged. You can ignore the failing buld job for Dashing, it is no longer relevant. Thanks for adding the test cases. Just added the test directly after the `remove`operation.

**Comment by BrettRD** [comment]
@JanStaschulat Thanks for the suggestions!

I didn't expect such a mess after the recommended signoff procedure!
That rebase made the files-changed tab almost unreadable

**Comment by JanStaschulat** [comment]
Hmm, something went wrong with updating DCO. Now it unfortunately contains many other files.

**Comment by BrettRD** [comment]
> Could you do a rebase on current Foxy branch first?

That looks *much* better.

**Review by JanStaschulat — APPROVED** [review]

**Comment by JanStaschulat** [comment]
@mergify backport galactic master

**Comment by JanStaschulat** [comment]
solves https://github.com/ros2/rclc/issues/50

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-humble-hawksbill-released/25729/14

**Pull request merged**
